shared/test/metabase/shared/models/visualization_settings_test.cljc

```diff
@@ -0,0 +1,152 @@
+  (t/testing ":visualization_settings are correctly converted from DB to qualified form and back"
+    (let [f-id                42
+          target-id           19
+          col-name            "My Column"
+          db-click-behavior   {:type             "link"
+                               :linkType         "question"
+                               :parameterMapping {}
+                               :targetId         target-id}
+          db-col-settings     {(fmt "[\"ref\",[\"field\",%d,null]]" f-id) {:click_behavior db-click-behavior}
+                               (fmt "[\"name\",\"%s\"]" col-name)         {:click_behavior db-click-behavior}}
+          db-viz-settings     {:column_settings db-col-settings}
+          norm-click-behavior {::mb.viz/click-behavior-type ::mb.viz/link
+                               ::mb.viz/link-type           ::mb.viz/card
+                               ::mb.viz/parameter-mapping   {}
+                               ::mb.viz/link-target-id      target-id}
+          norm-click-bhvr-map {::mb.viz/click-behavior norm-click-behavior}
+          norm-col-settings   {(mb.viz/column-ref-for-id f-id) norm-click-bhvr-map
+                               (mb.viz/column-ref-for-column-name col-name) norm-click-bhvr-map}
+          norm-viz-settings   {::mb.viz/column-settings norm-col-settings}]
+      (doseq [[db-form norm-form] [[db-viz-settings norm-viz-settings]]]
+        (let [to-norm (mb.viz/from-db-form db-form)]
+          (t/is (= norm-form to-norm))
+          (let [to-db (mb.viz/db-form to-norm)]
+            (t/is (= db-form to-db))))))))
+
+(t/deftest virtual-card-test
+  (t/testing "Virtual card in visualization settings is preserved through normalization roundtrip"
+    ;; virtual cards have the form of a regular card, mostly
+    (let [db-form {:virtual_card {:archived false
+                                  ;; the name is nil
+                                  :name     nil
+                                  ;; there is no dataset_query
+                                  :dataset_query {}
+                                  ;; display is text
+                                  :display "text"
+                                  ;; visualization settings also exist here (being a card), but are unused
+                                  :visualization_settings {}}
+                   ;; this is where the actual text is stored
+                   :text        "Stuff in Textbox"}]
+      ;; the current viz setting code does not interpret textbox type cards, hence this should be a passthrough
+      (t/is (= db-form (-> db-form
+                           mb.viz/from-db-form
+                           mb.viz/db-form))))))
+
+(t/deftest parameter-mapping-test
+  (t/testing "parameterMappings are handled correctly"
+    (let [from-id    101
+          to-id      294
+          card-id    19852
+          mapping-id (fmt "[\"dimension\",[\"fk->\",[\"field-id\",%d],[\"field-id\",%d]]]" from-id to-id)
```

Comment:
Legacy MBQL syntax.

Run it thru the normalizer and let that take care of it. That should be the only place in the whole codebase that knows about or deals with legacy syntax

Reply:
Fixed.